cpp/src/experimental/graph.cu

```diff
@@ -0,0 +1,543 @@
+        : partition_.vertex_partition_offsets[comm_p_col_size * (comm_p_row_rank + 1)];
+    auto col_first = partition_.vertex_partition_offsets[comm_p_row_size * comm_p_col_rank];
+    auto col_last  = partition_.vertex_partition_offsets[comm_p_row_size * (comm_p_col_rank + 1)];
+
+    rmm::device_uvector<edge_t> offsets(0, this->get_handle_ptr()->get_stream());
+    rmm::device_uvector<vertex_t> indices(0, this->get_handle_ptr()->get_stream());
+    rmm::device_uvector<weight_t> weights(0, this->get_handle_ptr()->get_stream());
+    std::tie(offsets, indices, weights) = edge_list_to_compressed_sparse<store_transposed>(
+      *(this->get_handle_ptr()), edgelists[i], row_first, row_last, col_first, col_last);
+    adj_matrix_partition_offsets_.push_back(std::move(offsets));
+    adj_matrix_partition_indices_.push_back(std::move(indices));
+    if (this->is_weighted()) { adj_matrix_partition_weights_.push_back(std::move(weights)); }
+  }
+
+  // update degree-based segment offsets (to be used for graph analytics kernel optimization)
+
+  if (sorted_by_global_degree_within_vertex_partition) {
+    auto degrees = detail::compute_major_degree(
+      *(this->get_handle_ptr()), adj_matrix_partition_offsets_, partition_);
+
+    // optinoal expensive checks (part 2/3)
+
+    if (do_expensive_check) {
+      CUGRAPH_EXPECTS(thrust::is_sorted(rmm::exec_policy(default_stream)->on(default_stream),
+                                        degrees.begin(),
+                                        degrees.end(),
+                                        thrust::greater<edge_t>{}),
+                      "Invalid API parameter: sorted_by_global_degree_within_vertex_partition is "
+                      "set to true, but degrees are not non-ascending.");
+    }
+
+    static_assert(detail::num_segments_per_vertex_partition == 3);
+    static_assert((detail::low_degree_threshold <= detail::mid_degree_threshold) &&
+                  (detail::mid_degree_threshold <= std::numeric_limits<edge_t>::max()));
+    rmm::device_uvector<edge_t> d_thresholds(detail::num_segments_per_vertex_partition - 1,
+                                             default_stream);
+    std::vector<edge_t> h_thresholds = {static_cast<edge_t>(detail::low_degree_threshold),
+                                        static_cast<edge_t>(detail::mid_degree_threshold)};
+    raft::update_device(
+      d_thresholds.data(), h_thresholds.data(), h_thresholds.size(), default_stream);
+
+    rmm::device_uvector<vertex_t> segment_offsets(detail::num_segments_per_vertex_partition + 1,
+                                                  default_stream);
+    segment_offsets.set_element_async(0, 0, default_stream);
+    segment_offsets.set_element_async(
+      detail::num_segments_per_vertex_partition, degrees.size(), default_stream);
+
+    thrust::upper_bound(rmm::exec_policy(default_stream)->on(default_stream),
```

Comment:
See CAVEAT: thrust lower_bound() / upper_bound() / binary_search(): thrust/thrust#1104 (although vectorized versions are okay, I think)

Reply:
Thanks but it seems like this does not apply to this case.